File: src/types/plan.ts
export type Currency = "CNY";
export type TransportMode = "subway" | "walk" | "taxi" | "drive" | "mixed";
export type CrowdLevel = "low" | "medium" | "high" | "unknown";

export type CardType = "transit" | "activity" | "dining" | "buffer";

export type PlanCardType =
  | "transport"
  | "activity"
  | "meal"
  | "rest"
  | "return"
  | "buffer";

export type CardStatus =
  | "done"
  | "active"
  | "current"
  | "upcoming"
  | "pending"
  | "risk"
  | "adjusted"
  | "skipped";

export type PlanStatus =
  | "draft"
  | "generating"
  | "ready"
  | "adjusting"
  | "confirmed"
  | "fallback";

export type PlanSource = "mock" | "api" | "llm" | "fallback";

export interface GeoPoint {
  lat: number;
  lng: number;
}

export interface Origin {
  name: string;
  address?: string;
  lat?: number;
  lng?: number;
}

export interface TimeWindow {
  start: string;
  end: string;
  mustReturnBy?: string;
}

export interface Participants {
  adults: number;
  children: number;
  childAges?: number[];
  pets?: number;
}

export interface BudgetConstraint {
  amount: number;
  currency: Currency;
  scope: "total" | "per_person";
}

export type SceneType =
  | "family"
  | "couple"
  | "friends"
  | "pet_friendly"
  | "indoor_rainy"
  | "low_queue"
  | "budget";

export interface ParsedConstraints {
  goalText: string;
  city?: string;
  origin: Origin;
  timeWindow: TimeWindow;
  participants: Participants;
  budget: BudgetConstraint;
  transportPreference: TransportMode;
  pacePreference: "relaxed" | "normal" | "packed";
  sceneType: SceneType;
  preferences: string[];
  avoidances: string[];
  weatherSensitive?: boolean;
  createdAt?: string;
}

export type LegacyTransportMode = string;
export type LegacyPace = string;

export interface LegacyConstraints {
  goal: string;
  time_start: string;
  time_end: string;
  adults: number;
  children: number;
  children_age: number;
  budget: number;
  departure: string;
  transport_mode: LegacyTransportMode;
  pace: LegacyPace;
  preference_tags: string[];
}

export interface Constraints {
  goal: string;
  goalText?: string;
  city?: string;
  origin?: Origin;
  time_start: string;
  time_end: string;
  timeWindow?: TimeWindow;
  adults: number;
  children: number;
  children_age: number;
  participants?: Participants;
  budget: number;
  budgetDetail?: BudgetConstraint;
  departure: string;
  transport_mode: LegacyTransportMode;
  transportPreference?: TransportMode;
  pace: LegacyPace;
  pacePreference?: "relaxed" | "normal" | "packed";
  sceneType?: SceneType;
  preference_tags: string[];
  preferences?: string[];
  avoidances?: string[];
  weatherSensitive?: boolean;
  createdAt?: string;
}

export type POIType =
  | "park"
  | "restaurant"
  | "cafe"
  | "mall"
  | "bookstore"
  | "activity"
  | "transport"
  | "home"
  | "other";

export type QueueLevel = CrowdLevel;

export type ReservationStatus =
  | "available"
  | "unavailable"
  | "waitlist"
  | "unknown";

export interface POI {
  id?: string;
  name: string;
  type?: POIType;
  categoryLabel?: string;
  address?: string;
  lat?: number;
  lng?: number;
  rating: number;
  avgPrice?: number;
  currency?: Currency;
  openingHours?: string;
  queueMinutes?: number;
  queueLevel?: QueueLevel;
  reservationStatus?: ReservationStatus;
  familyFriendly?: boolean;
  petFriendly?: boolean;
  indoor?: boolean;
  tags?: string[];

  poi_id: string;
  price_per_person: number;
  queue_minutes: number;
  category: string;
  map_position: { x: number; y: number };
  is_child_friendly: boolean;
  hours_label?: string;
  address?: string;
  district?: string;
  latitude?: number;
  longitude?: number;
  recommendation_reason?: string;
  risk_labels?: string[];
}

export interface RouteSegment {
  fromPoiId: string;
  toPoiId: string;
  mode: TransportMode;
  durationMinutes: number;
  distanceKm?: number;
  cost?: number;
  transferCount?: number;
  crowdLevel?: CrowdLevel;
  summary?: string;
}

export interface Card {
  id?: string;
  title?: string;
  subtitle?: string;
  planType?: PlanCardType;
  startTime?: string;
  endTime?: string;
  durationMinutes?: number;
  route?: RouteSegment;
  cost?: number;
  statusText?: string;
  reason?: string;
  adjustmentNote?: string;
  risks?: RiskSignal[];

  card_id: string;
  type: CardType;
  status: CardStatus;
  label: string;
  emoji: string;
  start_minute: number;
  duration_minutes: number;
  is_flexible: boolean;
  is_new?: boolean;
  poi?: POI;
  risk_note?: string;
  alternatives?: POI[];
}

export type PlanCard = Card;

export interface TimelineConfig {
  pixels_per_minute: number;
  min_card_width: number;
  card_height_map: {
    transit: number;
    activity: number;
    dining: number;
    buffer: number;
  };
}

export type RiskKind =
  | "queue"
  | "weather"
  | "budget"
  | "time"
  | "closure"
  | "fatigue";

export type ApiRiskKind =
  | "queue"
  | "weather"
  | "budget"
  | "time"
  | "traffic"
  | "closed"
  | "reservation"
  | "child_tired";

export interface RiskSignal {
  id?: string;
  type: RiskKind;
  level?: "low" | "medium" | "high";
  title: string;
  description: string;
  affectedCardIds?: string[];
  detectedAt?: string;
  suggestedAction?: string;
  requiresUserConfirm?: boolean;

  risk_id: string;
  severity: "medium" | "high";
  affected_card_ids: string[];
}

export interface PlanSummary {
  title: string;
  subtitle: string;
  totalBudget: number;
  budgetLimit: number;
  currency: Currency;
  totalDurationMinutes: number;
  expectedReturnTime: string;
  tags: string[];
  recommendationReason: string;
}

export interface PlanMapMarker {
  id: string;
  poiId?: string;
  cardId?: string;
  label: string;
  lat?: number;
  lng?: number;
  type: "current" | "next" | "normal" | "risk";
}

export interface PlanMap {
  currentCardId?: string;
  routePolyline?: GeoPoint[];
  markers?: PlanMapMarker[];
}

export interface PlanProgress {
  currentStep: number;
  totalSteps: number;
  percent: number;
  message: string;
}

export interface PlanDebugInfo {
  traceId?: string;
  llmUsed?: boolean;
  fallbackReason?: string;
}

export interface PlanBundle {
  planId?: string;
  version?: number;
  status?: PlanStatus;
  source?: PlanSource;
  constraints?: Constraints;
  summary?: PlanSummary;
  cards: Card[];
  risks?: RiskSignal[];
  map?: PlanMap;
  progress?: PlanProgress;
  updatedAt?: string;
  debug?: PlanDebugInfo;

  timeline: TimelineConfig;
}

export interface ParseGoalRequest {
  text: string;
  city?: string;
  originName?: string;
}

export interface ParseGoalResponse {
  constraints: Constraints;
  warnings?: string[];
}

export interface GeneratePlanRequest {
  constraints: Constraints;
}

export interface GeneratePlanResponse {
  plan: PlanBundle;
}

export type ReplanUserAction =
  | "accept_adjustment"
  | "reject_adjustment"
  | "regenerate"
  | "change_preference";

export interface ReplanRequest {
  plan: PlanBundle;
  risk?: RiskSignal;
  userAction?: ReplanUserAction;
  userMessage?: string;
}

export interface ReplanResponse {
  plan: PlanBundle;
  changedCardIds: string[];
  message: string;
}

export interface ConfirmPlanRequest {
  planId: string;
  plan: PlanBundle;
}

export interface ConfirmPlanResponse {
  success: boolean;
  planId: string;
  saved?: boolean;
  message?: string;
}

export type MachineState =
  | "IDLE"
  | "EXECUTING"
  | "RISK_DETECTED"
  | "REPLANNING"
  | "COMPLETED"
  | "CONFIRMED";

export interface PlanState {
  planId: string | null;
  version: number;
  cards: Card[];
  timeline: TimelineConfig;
  constraints: Constraints;
  planHistory: Card[][];
}

export type PlanAction =
  | { type: "SET_CARDS"; cards: Card[] }
  | { type: "SET_VERSION"; version: number }
  | { type: "PUSH_HISTORY" }
  | {
      type: "APPLY_CARD_PATCHES";
      patches: { card_id: string; patch: Partial<Card> }[];
    }
  | { type: "RESET_PLAN_DEMO"; cards: Card[] };

export interface AgentLogEntry {
  id: string;
  created_at: number;
  message: string;
}

export type ReplanPhase =
  | "idle"
  | "freezing"
  | "deconstructing"
  | "generating"
  | "animating"
  | "done";

export interface UIState {
  focusedCardId: string;
  activeRisk: RiskSignal | null;
  agentMessage: string | null;
  agentLogs: AgentLogEntry[];
  riskStatusSnapshot: Record<string, CardStatus> | null;
  replanPhase: ReplanPhase;
  replanInsertedOrder: string[] | null;
  autoRiskEnabled: boolean;
  riskAutoConsumed: boolean;
}

File: backend/app/schemas/plan.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

CardType = Literal["transit", "activity", "dining", "buffer"]
CardStatus = Literal["done", "active", "upcoming", "pending", "risk", "skipped"]
RiskKind = Literal["queue", "time", "budget", "closure", "weather", "fatigue"]
MachineState = Literal[
    "IDLE",
    "EXECUTING",
    "RISK_DETECTED",
    "REPLANNING",
    "COMPLETED",
    "CONFIRMED",
]


class CreatePlanRequest(BaseModel):
    prompt: str = Field(min_length=1, max_length=2000)
    city: str = "上海"
    timezone: str = "Asia/Shanghai"
    session_id: str | None = Field(default=None, max_length=128)
    user_id: str | None = Field(default=None, max_length=128)


class MapPosition(BaseModel):
    x: float
    y: float


class POI(BaseModel):
    poi_id: str
    name: str
    rating: float
    price_per_person: int
    queue_minutes: int
    category: str
    map_position: MapPosition
    is_child_friendly: bool
    hours_label: str | None = None
    address: str | None = None
    district: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    recommendation_reason: str | None = None
    risk_labels: list[str] = Field(default_factory=list)


class Card(BaseModel):
    card_id: str
    type: CardType
    status: CardStatus
    label: str
    emoji: str
    start_minute: int
    duration_minutes: int
    is_flexible: bool
    is_new: bool | None = None
    poi: POI | None = None
    risk_note: str | None = None
    alternatives: list[POI] | None = None


class Constraints(BaseModel):
    goal: str
    time_start: str
    time_end: str
    adults: int
    children: int
    children_age: int
    budget: int
    departure: str
    transport_mode: Literal["地铁", "驾车", "步行"]
    pace: Literal["轻松", "紧凑"]
    preference_tags: list[str]


class CardHeightMap(BaseModel):
    transit: int
    activity: int
    dining: int
    buffer: int


class TimelineConfig(BaseModel):
    pixels_per_minute: int
    min_card_width: int
    card_height_map: CardHeightMap


class RiskSignal(BaseModel):
    risk_id: str
    type: RiskKind
    severity: Literal["medium", "high"]
    title: str
    description: str
    affected_card_ids: list[str]


class AgentLogEntry(BaseModel):
    id: str
    created_at: int
    message: str


class PlanSummary(BaseModel):
    title: str
    subtitle: str


class PlanResponse(BaseModel):
    plan_id: str
    session_id: str
    user_id: str | None = None
    city: str = "上海"
    source: str = "api"
    status: MachineState
    version: int
    constraints: Constraints
    timeline: TimelineConfig
    cards: list[Card]
    active_risk: RiskSignal | None = None
    agent_logs: list[AgentLogEntry]
    summary: PlanSummary
    created_at: datetime
    updated_at: datetime


class PlanVersionResponse(BaseModel):
    version_id: str
    plan_id: str
    session_id: str
    user_id: str | None = None
    city: str = "上海"
    source: str = "api"
    version: int
    event_type: str
    status: MachineState
    constraints: Constraints
    timeline: TimelineConfig
    cards: list[Card]
    active_risk: RiskSignal | None = None
    agent_logs: list[AgentLogEntry]
    created_at: datetime


class PlanVersionCompareResponse(BaseModel):
    plan_id: str
    base_version_id: str
    target_version_id: str
    base_version: int
    target_version: int
    base_status: MachineState
    target_status: MachineState
    added_card_ids: list[str]
    removed_card_ids: list[str]
    changed_card_ids: list[str]
    unchanged_card_ids: list[str]
